File: sample/experiment-enums-compiler/java/src/main/java/com/uber/crumb/sample/experimentenumscompiler/ExperimentsCompiler.java
/*
 * Copyright (c) 2018. Uber Technologies
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.uber.crumb.sample.experimentenumscompiler;

import com.google.auto.common.MoreElements;
import com.google.auto.service.AutoService;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.ImmutableSet;
import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.CodeBlock;
import com.squareup.javapoet.FieldSpec;
import com.squareup.javapoet.JavaFile;
import com.squareup.javapoet.ParameterizedTypeName;
import com.squareup.javapoet.TypeName;
import com.squareup.javapoet.TypeSpec;
import com.uber.crumb.CrumbContext;
import com.uber.crumb.extensions.CrumbConsumerExtension;
import com.uber.crumb.extensions.CrumbProducerExtension;
import com.uber.crumb.sample.experimentsenumscompiler.annotations.Experiments;
import java.io.IOException;
import java.lang.annotation.Annotation;
import java.util.Arrays;
import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import javax.lang.model.element.AnnotationMirror;
import javax.lang.model.element.ElementKind;
import javax.lang.model.element.TypeElement;
import javax.tools.Diagnostic;

import static com.google.auto.common.MoreElements.asType;
import static java.util.stream.Collectors.toList;
import static java.util.stream.Collectors.toMap;
import static java.util.stream.Collectors.toSet;
import static javax.lang.model.element.Modifier.ABSTRACT;
import static javax.lang.model.element.Modifier.FINAL;
import static javax.lang.model.element.Modifier.PUBLIC;
import static javax.lang.model.element.Modifier.STATIC;

/**
 * A simple crumb producer/consumer that reads "experiment enums" from libraries and writes a
 * mapping record of them in a consuming app. Could be useful for tracking what experiments are on
 * the classpath.
 */
@AutoService({ CrumbProducerExtension.class, CrumbConsumerExtension.class })
public final class ExperimentsCompiler implements CrumbProducerExtension, CrumbConsumerExtension {

  private static final String EXPERIMENTS_NAME = Experiments.class.getCanonicalName();
  private static final String METADATA_KEY = "ExperimentsCompiler";

  @Override public boolean isConsumerApplicable(CrumbContext context,
      TypeElement type,
      Collection<? extends AnnotationMirror> annotations) {
    return isExperimentsAnnotationPresent(annotations);
  }

  @Override public boolean isProducerApplicable(CrumbContext context,
      TypeElement type,
      Collection<? extends AnnotationMirror> annotations) {
    return isExperimentsAnnotationPresent(annotations);
  }

  private boolean isExperimentsAnnotationPresent(Collection<? extends AnnotationMirror>
      annotations) {
    for (AnnotationMirror annotation : annotations) {
      TypeElement annotationTypeElement = asType(annotation.getAnnotationType()
          .asElement());
      if (annotationTypeElement.getQualifiedName()
          .contentEquals(EXPERIMENTS_NAME)) {
        return true;
      }
    }
    return false;
  }

  @Override public void consume(CrumbContext context,
      TypeElement type,
      Collection<? extends AnnotationMirror> annotations,
      Set<? extends Map<String, String>> metadata) {

    // Must be an abstract class because we're generating the backing implementation.
    if (type.getKind() != ElementKind.CLASS) {
      context.getProcessingEnv()
          .getMessager()
          .printMessage(Diagnostic.Kind.ERROR,
              "@"
                  + Experiments.class.getSimpleName()
                  + " is only applicable on classes when consuming!",
              type);
      return;
    } else if (!type.getModifiers()
        .contains(ABSTRACT)) {
      context.getProcessingEnv()
          .getMessager()
          .printMessage(Diagnostic.Kind.ERROR, "Must be abstract!", type);
    }

    // Map of enum TypeElement to its members
    Map<TypeElement, Set<String>> experimentClasses = metadata.stream()
        .map(data -> data.get(METADATA_KEY))
        .map(enumClass -> context.getProcessingEnv()
            .getElementUtils()
            .getTypeElement(enumClass))
        .collect(toMap(typeElement -> typeElement,
            typeElement -> typeElement.getEnclosedElements()
                .stream()
                .filter(e -> e.getKind() == ElementKind.ENUM_CONSTANT)
                .map(Object::toString)
                .collect(toSet())));

    FieldSpec mapField = FieldSpec.builder(ParameterizedTypeName.get(ClassName.get(Map.class),
        TypeName.get(Class.class),
        ParameterizedTypeName.get(List.class, String.class)), "EXPERIMENTS", PUBLIC, STATIC, FINAL)
        .initializer(CodeBlock.of("new $T<>()", LinkedHashMap.class))
        .build();

    CodeBlock.Builder staticInitBlock = CodeBlock.builder();
    experimentClasses.forEach((key, value) -> staticInitBlock.addStatement(
        "$N.put($T.class, $T.asList($L))",
        mapField,
        TypeName.get(key.asType()),
        TypeName.get(Arrays.class),
        String.join(", ",
            value.stream()
                .map(s -> "\"" + s + "\"")
                .collect(toList()))));

    TypeSpec generatedClass = TypeSpec.classBuilder("Experiments_" + type.getSimpleName()
        .toString())
        .addModifiers(PUBLIC, FINAL)
        .superclass(TypeName.get(type.asType()))
        .addField(mapField)
        .addStaticBlock(staticInitBlock.build())
        .build();

    try {
      JavaFile.builder(MoreElements.getPackage(type)
          .getQualifiedName()
          .toString(), generatedClass)
          .build()
          .writeTo(context.getProcessingEnv()
              .getFiler());
    } catch (IOException e) {
      context.getProcessingEnv()
          .getMessager()
          .printMessage(Diagnostic.Kind.ERROR,
              "Failed to write generated experiments mapping! " + e.getMessage(),
              type);
    }
  }

  @Override public Map<String, String> produce(CrumbContext context,
      TypeElement type,
      Collection<? extends AnnotationMirror> annotations) {
    if (type.getKind() != ElementKind.ENUM) {
      context.getProcessingEnv()
          .getMessager()
          .printMessage(Diagnostic.Kind.ERROR,
              "@"
                  + Experiments.class.getSimpleName()
                  + " is only applicable on enums when producing!",
              type);
      return ImmutableMap.of();
    }
    return ImmutableMap.of(METADATA_KEY,
        type.getQualifiedName()
            .toString());
  }

  @Override public Set<Class<? extends Annotation>> supportedConsumerAnnotations() {
    return ImmutableSet.of(Experiments.class);
  }

  @Override public Set<Class<? extends Annotation>> supportedProducerAnnotations() {
    return ImmutableSet.of(Experiments.class);
  }

  @Override public String key() {
    return METADATA_KEY;
  }
}

File: sample/experiment-enums-compiler/kotlin/src/main/kotlin/com/uber/crumb/sample/experimentenumscompiler/ExperimentsCompiler.kt
/*
 * Copyright (c) 2018. Uber Technologies
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.uber.crumb.sample.experimentenumscompiler

import com.google.auto.common.MoreElements.asType
import com.google.auto.service.AutoService
import com.squareup.kotlinpoet.FileSpec
import com.squareup.kotlinpoet.ParameterizedTypeName
import com.squareup.kotlinpoet.PropertySpec
import com.squareup.kotlinpoet.asClassName
import com.squareup.kotlinpoet.asTypeName
import com.uber.crumb.compiler.api.ConsumerMetadata
import com.uber.crumb.compiler.api.CrumbConsumerExtension
import com.uber.crumb.compiler.api.CrumbContext
import com.uber.crumb.compiler.api.CrumbProducerExtension
import com.uber.crumb.compiler.api.ExtensionKey
import com.uber.crumb.compiler.api.ProducerMetadata
import com.uber.crumb.sample.experimentsenumscompiler.annotations.Experiments
import me.eugeniomarletti.kotlin.metadata.KotlinClassMetadata
import me.eugeniomarletti.kotlin.metadata.classKind
import me.eugeniomarletti.kotlin.metadata.kaptGeneratedOption
import me.eugeniomarletti.kotlin.metadata.kotlinMetadata
import org.jetbrains.kotlin.serialization.ProtoBuf.Class.Kind
import java.io.File
import javax.lang.model.element.AnnotationMirror
import javax.lang.model.element.Element
import javax.lang.model.element.ElementKind.CLASS
import javax.lang.model.element.ElementKind.ENUM
import javax.lang.model.element.ElementKind.ENUM_CONSTANT
import javax.lang.model.element.TypeElement
import javax.tools.Diagnostic.Kind.ERROR

/**
 * A simple crumb producer/consumer that reads "experiment enums" from libraries and writes a
 * mapping record of them in a consuming app. Could be useful for tracking what experiments are on
 * the classpath.
 */
@AutoService(CrumbProducerExtension::class, CrumbConsumerExtension::class)
class ExperimentsCompiler : CrumbProducerExtension, CrumbConsumerExtension {

  companion object {
    private val EXPERIMENTS_NAME = Experiments::class.java.canonicalName
    private const val METADATA_KEY: ExtensionKey = "ExperimentsCompiler"
  }

  override fun isConsumerApplicable(context: CrumbContext,
      type: TypeElement,
      annotations: Collection<AnnotationMirror>): Boolean {
    return isExperimentsAnnotationPresent(annotations)
  }

  override fun isProducerApplicable(context: CrumbContext,
      type: TypeElement,
      annotations: Collection<AnnotationMirror>): Boolean {
    return isExperimentsAnnotationPresent(annotations)
  }

  private fun isExperimentsAnnotationPresent(annotations: Collection<AnnotationMirror>): Boolean {
    return annotations
        .asSequence()
        .map { asType(it.annotationType.asElement()) }
        .any { it.qualifiedName.contentEquals(EXPERIMENTS_NAME) }
  }

  override fun consume(context: CrumbContext,
      type: TypeElement,
      annotations: Collection<AnnotationMirror>,
      metadata: Set<ConsumerMetadata>) {

    // Must be a type that supports extension values
    if (type.kind != CLASS) {
      context.processingEnv
          .messager
          .printMessage(ERROR,
              "@${Experiments::class.java.simpleName} is only applicable on enums when consuming!",
              type)
      return
    }

    val kmetadata = type.kotlinMetadata

    if (kmetadata !is KotlinClassMetadata) {
      context.processingEnv
          .messager
          .printMessage(ERROR,
              "@${Experiments::class.java.simpleName} can't be applied to $type: must be a class. KMetadata was $kmetadata and annotations were [${type.annotationMirrors.joinToString { it.annotationType.asElement().simpleName }}]",
              type)
      return
    }

    val classData = kmetadata.data
    val (nameResolver, classProto) = classData

    // Must be an abstract class because we're generating the backing implementation.
    if (classProto.classKind != Kind.OBJECT) {
      context.processingEnv
          .messager
          .printMessage(ERROR,
              "@${Experiments::class.java.simpleName} can't be applied to $type: must be a Kotlin object class",
              type)
      return
    }

    val packageName = nameResolver.getString(classProto.fqName)
        .substringBeforeLast('/')
        .replace('/', '.')

    // Map of enum TypeElement to its members
    val experimentClasses = metadata
        .mapNotNull { it[METADATA_KEY] }
        .map { enumClass -> context.processingEnv.elementUtils.getTypeElement(enumClass) }
        .associate {
          it to it.enclosedElements
              .filter { it.kind == ENUM_CONSTANT }
              .map(Element::toString)
        }

    val initializerCode = experimentClasses
        .map { "%T::class.java to listOf(${it.value.joinToString(", ") { "\"%S\"" }})" }
        .joinToString(", ")
    val initializerValues = experimentClasses
        .keys
        .map { it.asClassName() }
    val mapProperty = PropertySpec.builder("EXPERIMENTS",
        ParameterizedTypeName.get(Map::class.asClassName(),
            Class::class.asTypeName(),
            ParameterizedTypeName.get(List::class.asClassName(),
                String::class.asTypeName())))
        .receiver(type.asClassName())
        .initializer("mapOf($initializerCode)", initializerValues)
        .build()

    // Generate the file
    val generatedDir = context.processingEnv.options[kaptGeneratedOption]?.let(::File)
        ?: throw IllegalStateException("Could not resolve kotlin generated directory!")
    FileSpec.builder(packageName, "${type.simpleName}_Experiments")
        .addProperty(mapProperty)
        .build()
        .writeTo(generatedDir)
  }

  override fun produce(context: CrumbContext,
      type: TypeElement,
      annotations: Collection<AnnotationMirror>): ProducerMetadata {
    if (type.kind != ENUM) {
      context.processingEnv
          .messager
          .printMessage(ERROR,
              "@${Experiments::class.java.simpleName} is only applicable on enums when producing!",
              type)
      return emptyMap()
    }
    return mapOf(METADATA_KEY to type.qualifiedName.toString())
  }

  override fun supportedConsumerAnnotations(): Set<Class<out Annotation>> = setOf(Experiments::class.java)

  override fun supportedProducerAnnotations(): Set<Class<out Annotation>> = setOf(Experiments::class.java)

  override fun key() = METADATA_KEY
}
